presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java

```diff
@@ -194,41 +206,83 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
         }
     }
 
-    @ThreadSafe
+    /**
+     * Each LookupSource has it's own copy of OuterPositionTracker instance.
+     * Each of those OuterPositionTracker must be committed after last write
+     * and before first read.
+     *
+     * All instances share visitedPositions array, but it is safe because each thread
+     * starts with visitedPositions filled with false values and marks only some positions
+     * to true. Since we don't care what will be the order of those writes to
+     * visitedPositions, writes can be without synchronization.
+     *
+     * Synchronization in commit() is required because each writing thread must have
```

Comment:
Does JMM say anything about Atomics explicitly?
Or just that write to volatile variable enforces happens-before semantic related to any subsequent read of same variable?

It is probably a safe assumption that AtomicLong uses volatile internally. Yet isn't that more explicit to just use monitor here which also enforces happens-before semantics between actions in threads entering it?

The correctness is then assured by the facts that:
1. All mutations to `vistitedPositions` are happening in each thread before this thread calls commit()
1. Each thread calls `commit()` after it performed all the mutations to `visitedPositions`
1. `getVisitedPositions` is called after all the threads which performed mutations to `visitedPositions` call `commit()`
1. commit() and `getVisitedPositions()` share monitor